src/target/adiv5_swd.c

```diff
@@ -79,133 +135,200 @@ static uint32_t firmware_dp_low_read(const uint16_t addr)
+ * - The host attempts auto-detection by issuing a target select command for each of the devices it has been
+ * configured to support.
+ * 
+ * This means that debug tools cannot connect seamlessly to targets in a multi-drop Serial Wire Debug system that they
+ * have never seen before. However, if the debug tools can be provided with the target ID of such targets by the user
+ * then the contents of the target can be auto-detected as normal.
+ * To protect against multiple selected devices all driving the line simultaneously SWD protocol version 2 requires:
+ * - For multi-drop SWJ-DP, the JTAG connection is selected out of powerup reset. JTAG does not drive the line.
+ * - For multi-drop SW-DP, the DP is in the dormant state out of powerup reset.
+ */
+void adiv5_swdp_multidrop_scan(adiv5_debug_port_s *target_dp, const uint32_t targetid)
+{
+	DEBUG_INFO("Handling swd multi-drop, TARGETID 0x%08" PRIx32 "\n", targetid);
 
-	const volatile size_t max_dp = scan_multidrop ? 16U : 1U;
-	for (volatile size_t i = 0; i < max_dp; i++) {
-		if (scan_multidrop) {
-			initial_dp->fault = 0;
-			swd_line_reset();
+	/* Scan all 16 possible instances (4-bit instance ID) */
+	for (size_t instance = 0; instance < 16U; instance++) {
+		/*
+		 * On a write to TARGETSEL immediately following a line reset sequence, the target is selected if both the following
+		 * conditions are met:
+		 * Bits [31:28] match bits [31:28] in the DLPIDR. (i.e. the instance ID matches)
+		 * Bits [27:0] match bits [27:0] in the TARGETID register.
+		 * Writing any other value deselects the target. 
+		 * During the response phase of a write to the TARGETSEL register, the target does not drive the line
+		 */
 
-			initial_dp->dp_low_write(ADIV5_DP_TARGETSEL,
-				i << ADIV5_DP_TARGETSEL_TINSTANCE_OFFSET |
-					(dp_targetid & (ADIV5_DP_TARGETID_TDESIGNER_MASK | ADIV5_DP_TARGETID_TPARTNO_MASK)) | 1U);
+		/* Line reset sequence */
+		swd_line_reset_sequence(true);
+		target_dp->fault = 0;
 
-			volatile uint32_t target_id = 0;
-			TRY_CATCH (e, EXCEPTION_ALL) {
-				target_id = adiv5_dp_low_access(initial_dp, ADIV5_LOW_READ, ADIV5_DP_DPIDR, 0U);
-			}
-			if (e.type || initial_dp->fault || !target_id)
-				continue;
-		}
+		/* Select the instance */
+		target_dp->dp_low_write(ADIV5_DP_TARGETSEL,
+			instance << ADIV5_DP_TARGETSEL_TINSTANCE_OFFSET |
+				(targetid & (ADIV5_DP_TARGETID_TDESIGNER_MASK | ADIV5_DP_TARGETID_TPARTNO_MASK)) | 1U);
 
-		adiv5_debug_port_s *dp = calloc(1, sizeof(*dp));
-		if (!dp) { /* calloc failed: heap exhaustion */
-			DEBUG_ERROR("calloc: failed in %s\n", __func__);
+		/* Read DPIDR */
+		if (adiv5_dp_read_dpidr(target_dp) == 0)
+			/* No DP here, next instance */
 			continue;
-		}
 
-		memcpy(dp, initial_dp, sizeof(adiv5_debug_port_s));
-		dp->instance = i;
+		/* Store the current instance DP, we will yield it to adiv5_dp_init and continue with a new one */
+		adiv5_debug_port_s *const instance_dp = target_dp;
```

Comment:
With the DP variable fed in renamed, this can become `target_dp = dp;` which will help the below code make more sense. We'd suggest keeping the original calloc-and-memcpy behaviour though as that's less confusing (less if logic and nesting) and allows the final action taken in this function to be `free(dp);` to keep things neat.

We know what you're doing here would result in less memory fragmentation, but the convolution of the resulting logic doesn't feel worth it.

Reply:
I was trying to be smart, but you're right it's a bit convoluted